cts/scheduler/exp/expired-stop-1.exp

```diff
@@ -1 +1,20 @@
-<transition_graph cluster-delay="60s" stonith-timeout="60s" failed-stop-offset="INFINITY" failed-start-offset="INFINITY"  transition_id="0"/>
+<transition_graph cluster-delay="60s" stonith-timeout="60s" failed-stop-offset="INFINITY" failed-start-offset="INFINITY"  transition_id="0">
+  <synapse id="0">
+    <action_set>
+      <rsc_op id="2" operation="stop" operation_key="rsc1_stop_0" on_node="node2" on_node_uuid="2">
+        <primitive id="rsc1" class="ocf" provider="pacemaker" type="Dummy"/>
+        <attributes CRM_meta_on_node="node2" CRM_meta_on_node_uuid="2" CRM_meta_timeout="20000" />
+      </rsc_op>
+    </action_set>
+    <inputs/>
+  </synapse>
+  <synapse id="1">
+    <action_set>
+      <crm_event id="1" operation="clear_failcount" operation_key="rsc1_clear_failcount_0" on_node="node2" on_node_uuid="2">
```

Comment:
If there are other failures that have not yet expired, will this clear the failcount that's due to those failures also?

Reply:
Yes, Pacemaker does not age out failed operations one by one as they pass the failure-timeout, but considers all failed operations for a given resource on a given node to be expired or unexpired together, based on the most recent one. Cleanup erases all fail-count and last-failure node attributes for the resource, as well as its RSC_last_failure_0 history entry.

Reply:
Makes sense. Makes me wonder if there are any edge cases for different operations. For example, if a recurring monitor can run and fail while a stop operation is expired, resetting the stop failure timeout. Or if there are two recurring monitors and each one's failure keeps resetting the other's failure timeout, so that the failures don't expire and we reach the migration threshold when we shouldn't.

But such cases wouldn't be new behavior and don't need to be addressed here.

Reply:
Yes, that is actually the intended behavior. It's not what a lot of people expect, but it's how Pacemaker defines expiration.

Reply:
A bit counterintuitive yeah. Looking at Pacemaker Explained again, it is documented, so that's good:

> The `migration-threshold` is per resource, even though fail counts are tracked per operation. The operation fail counts are added together to compare against the `migration-threshold`.
> ...
> `failure-timeout` is measured since the most recent failure. That is, older failures do not individually time out and lower the fail count. Instead, all failures are timed out simultaneously (and the fail count is reset to 0) if there is no new failure for the timeout period.

The behavior when multiple operations are failing is not obvious to me though, due to this clause: "fail counts are tracked per operation." This might suggest that the `failure-timeout` is **also** applied per operation. Sure, older failures don't individually time out and lower the fail count for Operation 1. But it would be reasonable to think that the most recent failure of Operation 1 could time out without being affected by a more recent failure of Operation 2.

Reply:
nothing is wrong, just a bit of gray area in that edge case